### PR TITLE
Customizable tunneling

### PR DESCRIPTION
**Pull request by Fefect**
Allows for custom tunnelling specified by given parameters.
The player location is the middle of the tunnel if possible (3 = uneven, possible: 4 = even, not possible, otherwise it'll stray to the right one block)
Example: https://dl.dropboxusercontent.com/s/yxya9u6nlyipzcy/2019-08-03_20-42-46.mp4

The current tunneling feature only supports a 2x1 area, this can support and width, height and depth in any of the 4 directions (north,east,south,west).

Code is still a bit dirty tho, not a big fan of Java overall.

### CODE REVIEW

**Comment by Fefect** [comment]
> 
> 
> Tunnel2?

Apologies, changed.

**Comment by StijnSimons** [comment]
.b is specified to impact. Not to baritone itself. Should be #

File: src/api/java/baritone/api/utils/ExampleBaritoneControl.java [hunk 1]
```diff
@@ -368,6 +369,82 @@ public boolean isInGoal(int x, int y, int z) {
             logDirect("tunneling");
             return true;
         }
+        if(msg.startsWith("tunnelcustom"))
```
Comment (ZeroMemes) [inline src/api/java/baritone/api/utils/ExampleBaritoneControl.java]: Formatting Standard: There should be a space between the `if` and the condition.

File: src/api/java/baritone/api/utils/ExampleBaritoneControl.java [hunk 2]
```diff
@@ -368,6 +369,82 @@ public boolean isInGoal(int x, int y, int z) {
...
+        {
```
Comment (ZeroMemes) [inline src/api/java/baritone/api/utils/ExampleBaritoneControl.java]: Formatting Standard: Opening braces should be on the line of some conditional, or expression. In this case: `if`.

File: src/api/java/baritone/api/utils/ExampleBaritoneControl.java [hunk 3]
```diff
@@ -368,6 +369,82 @@ public boolean isInGoal(int x, int y, int z) {
...
+                return true;
+            }
+            if(b < 2)
+            {
+                logDirect("Height must at least be 2 blocks");
+                return true;
+            }
+
+            //Adjustment for player location
+            a --;
```
Comment (ZeroMemes) [inline src/api/java/baritone/api/utils/ExampleBaritoneControl.java]: Formatting Standard: There shouldn't be a space between the variable identifier and inc/dec operators.

File: src/api/java/baritone/api/utils/ExampleBaritoneControl.java [hunk 4]
```diff
@@ -368,6 +369,82 @@ public boolean isInGoal(int x, int y, int z) {
...
+            //Check if width is uneven to center player's location
+            if(c % 2 ==0)
+            {
+                addition = 0;
+            }
+            switch(enumfacing)
+            {
+                case EAST:
+                    corner1 = new BlockPos(ctx.playerFeet().x , ctx.playerFeet().y , ctx.playerFeet().z - c/2);
+                    corner2 = new BlockPos(ctx.playerFeet().x + a, ctx.playerFeet().y + b, ctx.playerFeet().z + c/2 + addition);
```
Comment (ZeroMemes) [inline src/api/java/baritone/api/utils/ExampleBaritoneControl.java]: Formatting Standard: Mathematical operations should contain a single space between terms.

File: src/api/java/baritone/api/utils/ExampleBaritoneControl.java [hunk 5]
```diff
@@ -368,6 +369,82 @@ public boolean isInGoal(int x, int y, int z) {
...
             return true;
         }
+        if(msg.startsWith("tunnelcustom"))
+        {
+            String suffix = msg.substring("tunnelcustom".length());
+            int a;
+            int b;
+            int c;
+            if (suffix.isEmpty()) {
+                    logDirect("Need to specify depth, height and width of the tunnel. eg .b tunnel2 depth height width");
```
Comment (ZeroMemes) [inline src/api/java/baritone/api/utils/ExampleBaritoneControl.java]: `.b` is a command added by Impact *and* the command's name is `tunnelcustom`.

Should read something like `...eg #tunnel2 depth height width`

File: src/api/java/baritone/api/utils/ExampleBaritoneControl.java [hunk 6]
```diff
@@ -368,6 +369,82 @@ public boolean isInGoal(int x, int y, int z) {
...
+                    if(spl.length != 4) {
+                        logDirect(Integer.toString(spl.length));
+                        logDirect("Please enter 3 values");
+                        return true;
+                    }
+                    a = Integer.parseInt(spl[1]);
+                    b = Integer.parseInt(spl[2]);
+                    c = Integer.parseInt(spl[3]);
+                } catch (NumberFormatException | ArrayIndexOutOfBoundsException | NullPointerException ex) {
+                    logDirect("unable to parse");
```
Comment (ZeroMemes) [inline src/api/java/baritone/api/utils/ExampleBaritoneControl.java]: Should probably Make this error message `One or more number arguments were invalid` or something along the lines of this.

File: src/api/java/baritone/api/utils/ExampleBaritoneControl.java [hunk 7]
```diff
@@ -368,6 +369,82 @@ public boolean isInGoal(int x, int y, int z) {
...
+            //Adjustment for player location
+            a --;
+            b --;
+            c --;
+            BlockPos corner1 = ctx.playerFeet();
+            BlockPos corner2 = ctx.playerFeet();
+            EnumFacing enumfacing = ctx.player().getHorizontalFacing();
+            int addition = 1;
+            //Check if width is uneven to center player's location
+            if(c % 2 ==0)
```
Comment (ZeroMemes) [inline src/api/java/baritone/api/utils/ExampleBaritoneControl.java]: Formatting Specification: Boolean operators should contain spaces between its operands.

File: src/api/java/baritone/api/utils/ExampleBaritoneControl.java [hunk 8]
```diff
@@ -368,6 +369,82 @@ public boolean isInGoal(int x, int y, int z) {
...
+                    corner1 = new BlockPos(ctx.playerFeet().x , ctx.playerFeet().y , ctx.playerFeet().z + c/2 + addition);
+                    corner2 = new BlockPos(ctx.playerFeet().x - a, ctx.playerFeet().y + b, ctx.playerFeet().z - c / 2 );
+                    break;
+                case NORTH:
+                    corner1 = new BlockPos(ctx.playerFeet().x - c/2 , ctx.playerFeet().y , ctx.playerFeet().z);
+                    corner2 = new BlockPos(ctx.playerFeet().x + c/2 + addition, ctx.playerFeet().y + b, ctx.playerFeet().z - a);
+                    break;
+                case SOUTH:
+                    corner1 = new BlockPos(ctx.playerFeet().x + c/2 + addition, ctx.playerFeet().y , ctx.playerFeet().z);
+                    corner2 = new BlockPos(ctx.playerFeet().x - c/2 , ctx.playerFeet().y + b, ctx.playerFeet().z + a);
```
Comment (ZeroMemes) [inline src/api/java/baritone/api/utils/ExampleBaritoneControl.java]: Formatting Specification: The only spaces that should be immediately next to a comma when specifying arguments are succeeding spaces, not preceding.

**Comment by Fefect** [comment]
> 
> 
> A couple things I have initially noticed, otherwise this looks like it will be a good addition given the many requests to have it added.

Everything should be following the formatting rules now.

**Comment by StijnSimons** [comment]
S q u a s h

**Pull request closed without merging**